torch/_linalg_utils.py

```diff
@@ -0,0 +1,110 @@
+"""Various linear algebra utility methods for internal use.
+
+"""
+
+import torch
+
+
+def is_sparse(A):
+    """Check if tensor A is a sparse tensor"""
+    if isinstance(A, torch.Tensor):
+        return A.layout == torch.sparse_coo
+    raise TypeError("expected Tensor but got %s" % (type(A).__name__))
+
+
+def get_floating_dtype(A):
+    """Return the floating point dtype of tensor A.
+
+    Integer types map to float32.
+    """
+    dtype = A.dtype
+    if dtype in (torch.float16, torch.float32, torch.float64):
+        return dtype
+    return torch.float32
+
+
+def matmul(A, B):
+    """Multiply two matrices.
+
+    If A is None, return B. A can be sparse or dense. B is always
+    dense.
+    """
+    # type: (Optional[Tensor], Tensor) -> Tensor
+    if A is None:
+        return B
+    if is_sparse(A):
+        return torch.sparse.mm(A, B)
+    return torch.matmul(A, B)
+
+
+def conjugate(A):
+    """Return conjugate of tensor A.
+
+    .. note:: If A's dtype is not complex, A is returned.
+    """
+    if A.is_complex():
+        return A.conj()
+    return A
+
+
+def transpose(A):
+    """Return transpose of a matrix or batches of matrices.
+    """
+    ndim = len(A.shape)
+    return A.transpose(ndim - 1, ndim - 2)
+
+
+def transjugate(A):
+    """Return transpose conjugate of a matrix or batches of matrices.
+    """
+    return conjugate(transpose(A))
+
+
+def norm(A):
+    """Return Frobenius norm of a real matrix.
+    """
+    if A.is_cuda:
+        # e.g. conj is not available in CUDA
+        return matmul(transpose(A), A).trace() ** 0.5
+    return A.norm(2, [-2, -1])
+
```

Comment:
Frobenius norm is just `(A * A).sum()`. No need to do matrix products.

Reply:
Also, norm with parameter 2 computes Spectral norm, not Frobenium, am I wrong?

Reply:
good point re Fro norm calculations. Fixed.

`torch.norm` does not implement spectral norm. According to its docs, only the L(p,2) norms are supported.

Reply:
OK, but then `norm` is not supported for CUDA tensors?

Reply:
no, not at the time of introducing the code here: it failed with attribute error on `conj`.

Reply:
Peary, could you please explain why you use `**2` `**0.5`. Is it really as fast as just `x * x` or `sqrt`?

Reply:
I think `pow()` special-cases these, so there is no perf drop (cpu: https://github.com/pytorch/pytorch/blob/d8c368bd62455f2c9d0bf036640a87281db0face/aten/src/ATen/native/cpu/PowKernel.cpp#L42 cuda: https://github.com/pytorch/pytorch/blob/b565d9b35696259f9ab7d5dfd3814c82d40aaa84/aten/src/ATen/native/cuda/PowKernel.cu#L119)